### PR TITLE
Add _Float128 spelling for 128-bit float type

### PR DESCRIPTION
**Pull request by smowton**
According to https://gcc.gnu.org/onlinedocs/gcc/Floating-Types.html,
"On i386, x86_64, PowerPC, and IA-64 other than HP-UX, __float128 is an alias for _Float128"
and we support __float128, so we should probably support the other type name too.

As of glibc commit https://github.com/bminor/glibc/commit/fcee5905d341fe975f7786e72c831ada1c8fa78b, and in particular at least in the version shipped in Ubuntu 17.10, there are public `_Float128` declarations, which cause `goto-cc` to fail with a syntax error when including `math.h` due to the undefined type name.

Note this PR is probably not good enough yet as I don't really understand Flex/Bison, and just did the smallest hack to check this was the problem. I suggest someone who knows more about the parser and/or who knows more about what we intend to do with float128 should take it from here.

### CODE REVIEW

**Comment by smowton** [comment]
Note this is not reproducible with cbmc, presumably because it adds a bunch of custom defines before it runs the preprocessor which somehow prevent libc from exposing `_Float128`, whereas goto-cc runs plain old `gcc -E`

**Comment by kroening** [comment]
This looks fine. As a side remark, there's a plan to switch cbmc to have the same behaviour as goto-gcc.

**Review by kroening — APPROVED** [review]

**Comment by smowton** [comment]
One specific concern: what to do about https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/parser.y#L125 where it defines "__float128" as the canonical representation of the token. Is aliasing at the token level like this acceptable, or will it break some reverse-lookup routine?

**Comment by kroening** [comment]
It won't really break anything; it may look irritating when you get a parse error, and the token stream that gets printed has something in it you don't see in your source file. If you worry about this, you could add a second token (and the corresponding rules in the grammar). However, we have a couple of aliased tokens, and so far nobody has complained.

**Comment by smowton** [comment]
OK cool, in that case we just need tests I guess. I'll make a couple now.

**Comment by smowton** [comment]
Added to the existing type-compatibility test

File: regression/ansi-c/gcc_types_compatible_p1/main.c
Comment (tautschnig) [inline regression/ansi-c/gcc_types_compatible_p1/main.c]: This needs to be moved into the `#ifndef __clang__` block to get the regression test to pass on all platforms.

File: src/ansi-c/scanner.l
Comment (tautschnig) [inline src/ansi-c/scanner.l]: You could reduce the code duplication by adding that to the `"__float128"` block as `"__float128" | "_Float128"` as is done for, e.g., `"__complex__" | "__complex"`.

**Review by tautschnig — CHANGES_REQUESTED** [review]
The failing regression tests require correct preprocessor guards as indicated.

**Comment by smowton** [comment]
@tautschnig Done

**Comment by tautschnig** [comment]
Waiting for CI to complete, will merge once this has happened.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**
